metagraph/src/annotation/taxonomy/tax_classifier.hpp

```diff
@@ -0,0 +1,201 @@
+#ifndef __TAX_CLASSIFIER_HPP__
+#define __TAX_CLASSIFIER_HPP__
+
+#include <tsl/hopscotch_set.h>
+#include <tsl/hopscotch_map.h>
+
+#include "graph/annotated_dbg.hpp"
+
+namespace mtg {
+namespace annot {
+
+using TaxId = std::uint32_t;
+using ChildrenList = tsl::hopscotch_map<TaxId, std::vector<TaxId>>;
+
+class TaxonomyBase {
+  public:
+    using KmerId = annot::MultiLabelEncoded<std::string>::Index;
+    using node_index = graph::SequenceGraph::node_index;
+
+    enum LabelType {
+        UNASSIGNED,
+        GEN_BANK, // e.g. ">gi|1070643132|ref|NC_031224.1| Arthrobacter phage Mudcat, complete genome"
+        TAXID, // e.g. ">kraken:taxid|2016032|NC_047834.1 Alteromonas virus vB_AspP-H4/4, complete genome"
+    };
+
+    TaxonomyBase() {}
+    TaxonomyBase(double lca_coverage_rate, double kmers_discovery_rate)
+        : lca_coverage_rate_(lca_coverage_rate),
+          kmers_discovery_rate_(kmers_discovery_rate) {}
+    virtual ~TaxonomyBase() {}
+
+    TaxId assign_class(const std::string &sequence) const;
+
+  protected:
+    virtual TaxId find_lca(const std::vector<TaxId> &taxids) const = 0;
+
+    std::string get_accession_version_from_label(const std::string &label) const;
+
+    bool get_taxid_from_label(const std::string &label, TaxId *taxid) const;
+
+    /** Reads the accession version to taxid lookup table.
+     * If 'anno_matrix' is mentioned, only the labels that exist in the given annotation matrix will be stored.
+     * If 'anno_matrix' is not mentioned, the entire content of 'filepath' will be read and stored.
+     *
+     * @param [input] filepath -> a ".accession2taxid" file.
+     * @param [optional input] anno_matrix -> pointer to the annotation matrix.
+     */
+    void read_accversion_to_taxid_map(const std::string &filepath, const graph::AnnotatedDBG *anno_matrix);
+
+    /**
+     * Update the current node_scores and best_lca by taking into account the weight of the start_node and
+     * all its ancestors.
+     *
+     * @param [input] 'start_node' -> the starting node to update 'node_scores'.
+     * @param [input] 'num_kmers_per_node[taxid]' -> the number of kmers 'k' with taxonomic_map[k]=taxid.
+     * @param [input] 'desired_number_kmers' -> the threshold score that a node has to exceed in order to be
+     *                                          considered as a valid solution.
+     * @param [modified] 'node_scores' -> the current score for each node in the tree.
+     * @param [modified] 'nodes_already_propagated' -> the set of nodes that were previously processed.
+     * @param [modified] 'best_lca' -> the current classification prediction (node that exceeds the
+     *                                 `desired_number_kmers` threshold and is placed as close as possible to the leaves).
+     * @param [modified] 'best_lca_dist_to_root' -> the distance to the root for the current classification prediction.
+     */
+    void update_scores_and_lca(TaxId start_node,
```

Comment:
I don't see where this function is implemented or used. Remove it from this PR.
Also, check for this all other functions.

Reply:
sure, will remove them. I was just trying to present the complete structure of the objects: implement the entire `tax_classifier.hpp ` in this PR and not modify it in the next ones. But I see that it can be more clear to have the function header in the PR where it is implemented. I Will do the changes tomorrow, thanks!